### PR TITLE
Add OnQueryStarted hook

### PR DESCRIPTION
**Pull request by edevil**
Addresses #1043

### CODE REVIEW

**Comment by vmihailenco** [comment]
Hi,

Thanks for the PR and sorry for the delay.

I would expect that query started and query processed hooks receive the same event and you can save some data in query started hook event that is available in query processed hook. Also it probably makes sense to rename hooks like this:

```
type Hooks interface {
    OnBeforeQuery(*QueryEvent)
    OnAfterQuery(*QueryEvent)
}

type QueryEvent struct {
    // ... old fields here and
    Data map[interface{}]interface{} // here you can save some data
}

db.AddHook(&structImplementingHooksInterface{})
```

What do you think?

**Comment by edevil** [comment]
@vmihailenco That seems perfect for me, but that would be a breaking change. Is that Ok?

**Comment by vmihailenco** [comment]
Yes, I think that is fine as long as we push new major version.

**Comment by edevil** [comment]
@vmihailenco done, PTAL

**Comment by edevil** [comment]
@vmihailenco ping :)

File: hook.go [hunk 1]
```diff
@@ -62,43 +69,58 @@ func queryString(query interface{}) ([]byte, error) {
...
-	event := &QueryProcessedEvent{
-		StartTime: start,
-		Func:      funcName,
-		File:      file,
-		Line:      line,
+	event := &QueryEvent{
+		StartTime: time.Now(),
+		Func: funcName,
+		File: file,
+		Line: line,
```
Comment (vmihailenco) [inline hook.go]: Can you please remove these params? I think it is reasonable to expect that users should save info they need in `Data` map.

**Comment by vmihailenco** [comment]
I like it! One small change and this should be good to go.

File: hook.go (hunk 1 above)
Comment (edevil) [reply hook.go]: Done.

**Comment by edevil** [comment]
@vmihailenco Implemented the change you suggested. PTAL :)

File: hook.go [hunk 2]
```diff
@@ -62,84 +67,51 @@ func queryString(query interface{}) ([]byte, error) {
...
 }
 
-type queryProcessedHook func(*QueryProcessedEvent)
-
-// OnQueryProcessed calls the fn with QueryProcessedEvent
+// OnQueryEvent calls the interface with QueryEvent
 // when query is processed.
-func (db *DB) OnQueryProcessed(fn func(*QueryProcessedEvent)) {
-	db.queryProcessedHooks = append(db.queryProcessedHooks, fn)
+func (db *DB) OnQueryEvent(hook Hook) {
```
Comment (vmihailenco) [inline hook.go]: I will change this myself, but I think this should be `AddHook` or `AddQueryHook`. I expect `OnSomething` to receive a function.

**Comment by vmihailenco** [comment]
Thanks! Looks pretty good. Let me know if you spot that anything is still missing.

File: hook.go (hunk 2 above)
Comment (edevil) [reply hook.go]: Yeah, I didn’t give the name much thought, I agree.

**Comment by edevil** [comment]
@MrBadge You can store `time.Now()` on the `Data` map in "BeforeQuery", and use it on "AfterQuery".

**Comment by MrBadge** [comment]
> @MrBadge You can store `time.Now()` on the `Data` map in "BeforeQuery", and use it on "AfterQuery".

Thanks, that's indeed more convenient

**Pull request merged**